Use attribute name from config on LDAP group creation

jntesteves opened the pull request:
Use CommonLDAPGroupMapperConfig.getMembershipLdapAttribute() instead of
constant LDAPConstants.MEMBER to honor the "membership.ldap.attribute"
config key when creating a LDAP group. This fixes an error when trying
to create a group on a DS server configured with a different member
attribute than the standard "member" (eg. 389ds).

Fixes issue:
https://issues.jboss.org/browse/KEYCLOAK-7990
group-ldap-mapper setting "Membership LDAP Attribute" is not honored when creating group. The default value "member" is always used.

jntesteves committed:
KEYCLOAK-7990 Use attribute name from config on LDAP group creation

Use CommonLDAPGroupMapperConfig.getMembershipLdapAttribute() instead of
constant LDAPConstants.MEMBER to honor the "membership.ldap.attribute"
config key when creating a LDAP group. This fixes an error when trying
to create a group on a DS server configured with a different member
attribute than the standard "member" (eg. 389ds).

jntesteves commented:
Hello @mposolda, this is my new implementation for KEYCLOAK-7990. I followed your suggestions, and in the process I changed some things to make then more readable, I wasn't comfortable with removing those lines from a method I couldn't quite understand what it was doing.

I didn't implement any automated tests yet. I don't usually work with java and I'm not very acquainted with the testing frameworks. Are there any tests already for LDAP? Do we run a DS Server for the tests?

jntesteves commented:
Hi @mposolda, just a follow up to my last comment, I found the test suit and the LDAP tests and got to understand how it all works, I think. I've started doing the tests but due to other things coming up both at work and life, I'm currently out of time to finish it. I'll update this PR when I get to finish it.

Or, if you think this PR looks good and this change is urgent enough to merge now, I'll happily open a new PR with the finished tests later, I promise :grin:

mposolda commented:
@jntesteves Thanks for the PR! It looks good to me. Sorry for longer review. Just one small change, is it possible to add "KEYCLOAK-7990 " to the beginning of the commit message? See the other commits on Keycloak project.

jntesteves commented:
Hi, just changed the commit message, and rebased on master. I force-pushed the same branch. I don't know if that's the standard procedure, but I don't know any other way to update a github PR.

mposolda commented:
@jntesteves That is fine. Thanks for the update!

mposolda commented:
@jntesteves Sorry, one more thing. 

I think that PR will probably break MS Active Directory. The reason is, that on MSAD you can't create group with empty membership attribute. It differs from OpenLDAP and some other LDAP servers where the empty "member" attribute is required, but on the MSAD it is the exact opposite and the empty "member" attribute MUST NOT be there.

The check for group classes, which you removed, still must be there. MSAD uses the object class "group" and hence empty "member" won't be added for it, which is fine. In other words, can you please change the code in LDAPUtils.createLDAPGroup like this (please include the comment as well for future reference):
````
// On MSAD with object class "group", empty member must not be added. Specified object classes typically 
// require empty member attribute
if (objectClassValue.equalsIgnoreCase(LDAPConstants.GROUP_OF_NAMES)
        || objectClassValue.equalsIgnoreCase(LDAPConstants.GROUP_OF_ENTRIES)
        || objectClassValue.equalsIgnoreCase(LDAPConstants.GROUP_OF_UNIQUE_NAMES)) {
    ldapObject.setSingleAttribute(membershipLdapAttribute, LDAPConstants.EMPTY_MEMBER_ATTRIBUTE_VALUE)                   
}
````

jntesteves commented:
Ok, I changed that. I'm building a new version of our fork to test this more carefully on our environments and later I'll push here.

jntesteves commented:
@mposolda Updated, all seems OK. Thanks for the review and explanation.

mposolda commented:
@jntesteves Thanks!
